src/RADOS/cls/fifo.h

```diff
@@ -0,0 +1,1575 @@
+	} else  {
+	  batch.erase(batch.begin(), batch.begin() + successes);
+	  for (const auto& b : batch) {
+	    batch_len +=  b.length() + part_entry_overhead;
+	  }
+	}
+      }
+
+      if (batch.empty() && remaining.empty()) {
+	std::move(handler)(bs::error_code{});
+	return;
+      }
+
+      while (!remaining.empty() &&
+	     (remaining.front().length() + batch_len <= max_part_size)) {
+
+	/* We can send entries with data_len up to max_entry_size,
+	   however, we want to also account the overhead when
+	   dealing with multiple entries. Previous check doesn't
+	   account for overhead on purpose. */
+	batch_len += remaining.front().length() + part_entry_overhead;
+	batch.push_back(std::move(remaining.front()));
+	remaining.pop_front();
+      }
+      push();
+    }
+
+    void push() {
+      auto e = ba::get_associated_executor(handler);
+      auto a = ba::get_associated_allocator(handler);
+      f->push_entries(batch,
+		      ca::bind_ea(e, a,
+				  Pusher(f, std::move(remaining),
+					 batch, i,
+					 std::move(handler))));
+    }
+
+  public:
+
+    // Initial call!
+    void operator ()() {
+      prep_then_push(0);
+    }
+
+    // Called with response to push_entries
+    void operator ()(bs::error_code ec, int r) {
+      if (ec == bs::errc::result_out_of_range) {
+	auto e = ba::get_associated_executor(handler);
+	auto a = ba::get_associated_allocator(handler);
+	f->_prepare_new_head(
+	  ca::bind_ea(e, a,
+		      Pusher(f, std::move(remaining),
+			     std::move(batch), i,
+			     std::move(handler))));
+	return;
+      }
+      if (ec) {
+	std::move(handler)(ec);
+	return;
+      }
+      i = 0; // We've made forward progress, so reset the race counter!
+      prep_then_push(r);
+    }
+
+    // Called with response to prepare_new_head
+    void operator ()(bs::error_code ec) {
+      if (ec == bs::errc::operation_canceled) {
+	if (i == MAX_RACE_RETRIES) {
+	  ldout(f->r->cct(), 0)
+	    << "ERROR: " << __func__
+	    << "(): race check failed too many times, likely a bug" << dendl;
+	  std::move(handler)(make_error_code(errc::raced));
+	  return;
+	}
+	++i;
+      } else if (ec) {
+	std::move(handler)(ec);
+	return;
+      }
+
+      if (batch.empty()) {
+	prep_then_push(0);
+	return;
+      } else {
+	push();
+	return;
+      }
+    }
+
+    Pusher(FIFO* f, std::deque<cb::list>&& remaining,
+	   std::deque<cb::list> batch, int i,
+	   Handler&& handler)
+      : f(f), remaining(std::move(remaining)),
+	batch(std::move(batch)), i(i),
+	handler(std::move(handler)) {}
+  };
+
+  template<typename CT>
+  auto push(const std::vector<cb::list>& data_bufs,
+	    CT&& ct) {
+    ba::async_completion<CT, void(bs::error_code)> init(ct);
+    std::unique_lock l(m);
+    auto max_entry_size = info.params.max_entry_size;
+    auto need_new_head = info.need_new_head();
+    l.unlock();
+    if (data_bufs.empty() ) {
+      // Can't fail if you don't try.
+      std::move(init.completion_handler)({});
+      return init.result.get();
```

Comment:
I'm having a problem with this. If I use `ex2.dispatch(std::move(f), alloc2)` or `ba::dispatch(ex2, std::move(f))` the program crashes with

```
ceph_test_cls_fifo: /home/azure/programming/ceph/build/boost/include/boost/coroutine/detail/push_coroutine_impl.hpp:252: void boost::coroutines::detail::push_coroutine_impl<void>::unwind_stack(): Assertion `is_complete()' failed.`
```
This is very strange since using just `ba::dispatch(std::move(f))` works fine. As the free function is documented to call the associated executor (and in this case `ex2` was got with `ba::get_associated_executor`), I'm not sure why it should be any different than the other two.

Reply:
something was trying to use this default-constructed executor. it doesn't crash without this change
```diff
diff --git a/src/neorados/cls/fifo.h b/src/neorados/cls/fifo.h
index 7e68925c9b..5e00d42f93 100644
--- a/src/neorados/cls/fifo.h
+++ b/src/neorados/cls/fifo.h
@@ -198,7 +198,6 @@ class FIFO {
 
   template<typename Handler>
   class MetaReader {
-    ba::executor executor;
     Handler handler;
     using allocator_type = boost::asio::associated_allocator_t<Handler>;
     using decoder_type = ExecDecodeCB<fifo::op::get_meta_reply>;
@@ -215,7 +214,7 @@ class FIFO {
       auto reply = std::move(decoder->result);
       decoder.reset(); // free handler-allocated memory before dispatching
 
-      auto e = ba::get_associated_executor(handler, executor);
+      auto e = ba::get_associated_executor(handler);
       auto a = ba::get_associated_allocator(handler);
       auto p = [ec = ec, handler = std::move(handler),
                reply = std::move(reply)]() mutable {
```
for these Handler wrappers, `operator()()` is already running inside of its executor, so it _is_ okay to call the wrapped handler directly!

in this `read_meta()` case, you passed this Handler to `RADOS::execute()` - and when _that_ operation completes in Objecter, `ceph::async::Completion::defer()` is what submits it to the correct executor